src/openfermion/ops/_symbolic_operator.py

```diff
@@ -0,0 +1,520 @@
+    of the form (`index`, `action`), where `index` is a nonnegative integer
+    and the possible values for `action` are determined by the subclass.
+    For instance, for the subclass FermionOperator, `action` can be 1 or 0,
+    indicating raising or lowering, and for QubitOperator, `action` is from
+    the set {'X', 'Y', 'Z'}.
+    The coefficients of the terms are stored in a dictionary whose
+    keys are the terms.
+    SymbolicOperators of the same type can be added or multiplied together.
+
+    Attributes:
+        actions (tuple): A tuple of objects representing the possible actions.
+            This should be defined in the subclass.
+            e.g. for FermionOperator, this is (1, 0).
+        action_strings (tuple): A tuple of string representations of actions.
+            These should be in one-to-one correspondence with actions and
+            listed in the same order.
+            e.g. for FermionOperator, this is ('^', '').
+        action_before_index (bool): A boolean indicating whether in string
+            representations, the action should come before the index.
+        different_indices_commute (bool): A boolean indicating whether
+            factors acting on different indices commute.
+        terms (dict):
+            **key** (tuple of tuples): A dictionary storing the coefficients
+            of the terms in the operator. The keys are the terms.
+            A term is a product of individual factors; each factor is
+            represented by a tuple of the form (`index`, `action`), and
+            these tuples are collected into a larger tuple which represents
+            the term as the product of its factors.
+    """
+    actions = ()
+    action_strings = ()
+    action_before_index = False
+    different_indices_commute = False
+
+    def __init__(self, term=None, coefficient=1.):
+        if not isinstance(coefficient, (int, float, complex)):
+            raise ValueError('Coefficient must be a numeric type.')
+
+        # Detect if the input is the string representation of a sum of terms;
+        # if so, initialization needs to be handled differently
+        if isinstance(term, str) and '+' in term:
+            self._long_string_init(term, coefficient)
+            return
+
+        # Initialize the terms dictionary
+        self.terms = {}
+
+        # Zero operator: leave the terms dictionary empty
+        if term is None:
+            return
+
+        # Parse the term
+        # Sequence input
+        elif isinstance(term, tuple) or isinstance(term, list):
+            term = self._parse_sequence(term)
+        # String input
+        elif isinstance(term, str):
+            term = self._parse_string(term)
+        # Invalid input type
+        else:
+            raise ValueError('term specified incorrectly.')
+
+        # Add the term to the dictionary
+        self.terms[term] = coefficient
+
+    def _long_string_init(self, term, coefficient):
```

Comment:
Is there a reason why this needs to be implemented differently for FermionOperator and QubitOperator? Otherwise, I feel we should implement the method now.

Reply:
No, there is no reason. It's just that @max-radin said he would do it at #177 so I figured I'd leave it to him. Perhaps we should post there asking if he's still willing to do it?

Reply:
Yeah let's ask him about that.